SI-8015 Count lines by EOLs

som-snytt opened the pull request:
Source lines were counted by "line break chars", including FF.
Clients of `pos.line` seem to all expect the ordinary line num,
so that is what they get.

Unicode processing now precedes line ending processing.

Review by @adriaanm who is going to come across it anyway in the final hours of clean-up before cutting a milestone

som-snytt committed:
SI-8015 Count lines by EOLs

Source lines were counted by "line break chars", including FF.
Clients of `pos.line` seem to all expect the ordinary line num,
so that is what they get.

Unicode processing now precedes line ending processing.

som-snytt committed:
SI-8015 Unprintables in messages

The line content to display should include the whole line,
inclusive of embedded line breaks, and any control chars
should be displayed in unicode escape.

som-snytt committed:
SI-8015 Carat => Caret

Just a diamond in the rough.

Wikipedia warns against confusing it with "carrot."

The first word should really be pronounced as the garage phrase,
"Where's the car at?" and the second like "beret."

som-snytt commented:
Review by @paulp of the third commit.

It occurred to me that maybe a caret looks like half a diamond, and since the caret often indicates only roughly where an error is located, the epithet "diamond in the rough" is apt.  I don't know if it's in the style guide, but pun consistency is more important than correctness.

retronym commented:
This patch needs a rebase.

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:
```diff
@@ -389,7 +389,7 @@ trait Scanners extends ScannersCommon {
               return true
             }
-	    if (idx == end) return false
+            if (idx == end) return false
```
Not sure fixing this tab (which I recall really did cause a moment of confusion due to the do-while preceding it) was worth incurring the commit collision and rebase at line 210.

som-snytt commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -202,12 +202,23 @@ private[util] trait InternalPositionImpl {
   def lineContent: String = if (hasSource) source.lineToString(line - 1) else ""
-  def lineCarat: String   = if (hasSource) " " * (column - 1) + "^" else ""
+  def lineCaret: String   = if (hasSource) " " * (column - 1) + "^" else ""
+  @deprecated("use `lineCaret`", since="2.11.0")
```
If you swap the args, you'll get a warning.  We compile `-Xfatal-warnings`, of course.  Anyway, without the named arg, you have to rely on the Deprecator plugin to check that your version strings look like versions (and you didn't swap the unnamed args).

retronym commented:
@densh, could you please take a look at this one?

densh commented:
@retronym Sure, I'll review it on Monday.

paulp commented:
YER ALL MISPEWLING YER CAR ROT LMGT

densh commented on test/files/neg/t8015-ffb.scala:
```diff
@@ -0,0 +1,11 @@
+
+trait G {
+  val c: Char = '\u000a'   // disallowed!
```
I guess we have one more contradictory point in the spec now, don't we?

```
Note that ‘\u000A’ is not a valid character literal because Unicode conversion is done
before literal parsing and the Unicode character \u000A (line feed) is not a printable
character. One can use instead the escape sequence ‘\n’ or the octal escape ‘\12’ (§1.3.6).
```

som-snytt replied:
I asked on the ML but no one was shocked into action. It's too bad code needs line endings at all.  Why can't code just be a never-ending waveform, maybe pulsating a bit so you get a feel for the rhythm of it.

adriaanm replied:
And we could use winamp plugins for code review.

densh commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -202,12 +202,23 @@ private[util] trait InternalPositionImpl {
-    case pos           => s"${pos.line}: $msg\n${pos.lineContent}\n${pos.lineCarat}"
+    case pos           => f"${pos.line}: $msg%n${u(pos.lineContent)}%n${pos.lineCaret}"
+  }
+  private def u(s: String) = {
```
This method can definitely use a longer descriptive name.

densh commented on src/reflect/scala/reflect/internal/util/SourceFile.scala:
```diff
@@ -16,12 +16,13 @@ import scala.reflect.internal.Chars._
+  def content: Array[Char]         // normalized, must end in SU
+  def file   : AbstractFile
+  def isLineBreak(idx: Int): Boolean
+  def isEndOfLine(idx: Int): Boolean
```
Do we need to have both `isLineBreak` and `isEndOfLine`, can we get away with just one?

som-snytt replied:
We don't, but someone may.  Mima and downstream projects make me shy of changing API.  Since they're not protected final, it could be overridden; SourceFile is not the most internal of the internal API.  The impls could be DRYer, however, as I felt at the time.

adriaanm commented:
To someone not very spun up on scanner matters, looks good, but I'll defer to the expert @densh.

som-snytt committed:
SI-8015 Refactor per code review

Make an obscure private def local and rename it usefully.

If you really must have similar functions in the API
(to detect line breaks and EOL), at least make them DRY.

som-snytt commented:
I pushed a clean-up commit.  I realize I'll be spending some time in purgatory for all that I have done and left undone.  I figure I can get caught up on my reading.

som-snytt commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -202,12 +202,31 @@ private[util] trait InternalPositionImpl {
+      def where     = p.line
+      def content   = escaped(p.lineContent)
+      def indicator = p.lineCaret
+      f"$where: $msg%n$content%n$indicator"
```
It seems like such a little thing, but it really matters how you express an interpolated string.  I realized this after looking at this delta again after it sat on the PR queue for a month.  (It was only since last year.)

A few pointy braces and you want to go back to Dr Odersky's "simple"+ rule.

Edit: I meant the stradd rule, just keep the plus next to the quote and it's easy to read and write.

densh commented:
LGTM
